File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] arr) {
        int n = arr.length;
        int[] leftMax = new int[n];
        int[] rightMax = new int[n];
        
        leftMax[0] = arr[0];
        
        for(int i = 1; i < n; i++){
            leftMax[i] = Math.max(leftMax[i - 1], arr[i]);
        }
        
        rightMax[n - 1] = arr[n - 1];
        
        for(int i = n - 2; i >= 0; i--){
             rightMax[i] = Math.max(rightMax[i + 1], arr[i]);
        }
        
        int sum = 0;
        
        for(int i = 0; i < n; i++){
            sum += (Math.min(leftMax[i], rightMax[i]) - arr[i]);
        }
        
        return sum;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution
{
    public:
        int trap(vector<int> &height)
        {
            vector<int> leftMax(height.size());
            vector<int> rightMax(height.size());
            int l = height[0], r = height[height.size() - 1];
            leftMax[0] = l;
            rightMax[rightMax.size() - 1] = r;
            for (int i = 1; i < height.size(); i++)
            {
                l = max(l, height[i]);
                leftMax[i] = l;
            }
            for (int i = height.size() - 2; i >= 0; i--) {
                r = max(r, height[i]);
                rightMax[i] = r;
            }
            int res = 0;
            for(int i=0;i<height.size();i++){
                res = res + min(leftMax[i], rightMax[i]) - height[i];
            }
            return res;
        }
};